google-cloud-spanner/src/test/java/com/google/cloud/spanner/MultiplexedSessionDatabaseClientMockServerTest.java

```diff
@@ -1290,6 +1291,231 @@ public void testMutationOnlyUsingAsyncTransactionManager() {
+    DatabaseClientImpl client =
+        (DatabaseClientImpl) spanner.getDatabaseClient(DatabaseId.of("p", "i", "d"));
+
+    // Get the runner so that the read-write transaction is executed via multiplexed session.
+    TransactionRunner runner = client.readWriteTransaction();
+
+    // Unfreeze the mock server to get the error from the backend. The above read-write transaction
+    // will then fail.
+    mockSpanner.unfreeze();
+
+    SpannerException e =
+        assertThrows(
+            SpannerException.class,
+            () ->
+                runner.run(
+                    transaction -> {
+                      ResultSet resultSet = transaction.executeQuery(STATEMENT);
+                      //noinspection StatementWithEmptyBody
+                      while (resultSet.next()) {
+                        // ignore
+                      }
+                      return null;
+                    }));
+    assertEquals(ErrorCode.UNIMPLEMENTED, e.getErrorCode());
+
+    // Wait until the client sees that MultiplexedSessions are not supported for read-write.
+    assertNotNull(client.multiplexedSessionDatabaseClient);
+    SpannerException spannerException =
+        assertThrows(
+            SpannerException.class,
+            client.multiplexedSessionDatabaseClient::getReadWriteBeginTransactionReference);
+    assertEquals(ErrorCode.UNIMPLEMENTED, spannerException.getErrorCode());
+    assertTrue(client.multiplexedSessionDatabaseClient.unimplementedForRW.get());
+
+    // The next read-write transaction will fall back to regular sessions and succeed.
+    client
+        .readWriteTransaction()
+        .run(
+            transaction -> {
+              try (ResultSet resultSet = transaction.executeQuery(STATEMENT)) {
+                //noinspection StatementWithEmptyBody
+                while (resultSet.next()) {
+                  // ignore
+                }
+              }
+              return null;
+            });
+
+    // Verify that two ExecuteSqlRequests were received: the first using a multiplexed session and
+    // the second using a regular session.
+    assertEquals(2, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+    List<ExecuteSqlRequest> requests = mockSpanner.getRequestsOfType(ExecuteSqlRequest.class);
+
+    Session session1 = mockSpanner.getSession(requests.get(0).getSession());
+    assertNotNull(session1);
+    assertTrue(session1.getMultiplexed());
+
+    Session session2 = mockSpanner.getSession(requests.get(1).getSession());
+    assertNotNull(session2);
+    assertFalse(session2.getMultiplexed());
+
+    assertNotNull(client.multiplexedSessionDatabaseClient);
+    assertEquals(1L, client.multiplexedSessionDatabaseClient.getNumSessionsAcquired().get());
+    assertEquals(1L, client.multiplexedSessionDatabaseClient.getNumSessionsReleased().get());
+  }
+
+  @Test
+  public void testReadWriteUnimplemented_firstReceivesError_secondFallsBackToRegularSessions() {
```

Comment:
Can we add a test that also verifies that multiplexed sessions continues to work for read/write transactions after some other operation (e.g. a query using a single-use read-only transaction) returned UNIMPLEMENTED?

Reply:
Added